caom2pipe/execute_composable.py

```diff
@@ -200,182 +186,63 @@ def __init__(
-        # so far, the plugin is also the module :)
-        temp = (
-            f'{self.command_name} {self.logging_level_param} '
-            f'{self.cred_param} --in {self.model_fqn} --out {self.model_fqn} '
-            f'{self.external_urls_param} --plugin {plugin} --module '
-            f'{plugin} --lineage {self._storage_name.lineage} '
-            f'{self._find_fits2caom2_resource_id()}'
-        )
-        self._invoke_to_caom2(temp, plugin)
-
-    def _fits2caom2_cmd_in_out_local(self, connected=True):
-        """Execute fits2caom with a --in, --local and a --cert parameter."""
-        plugin = self._find_fits2caom2_plugin()
-        # so far, the plugin is also the module :)
-        local_fqn = ' '.join(ii for ii in self._storage_name.source_names)
-        conn = ''
-        resource_id = ''
-        if connected:
-            if self.supports_latest_client:
-                resource_id = (
-                    f'--resource-id {self._storage_inventory_resource_id}'
-                )
-        else:
-            conn = f'--not_connected'
-        temp = (
-            f'{self.command_name} {self.logging_level_param} {conn} '
-            f'{self.cred_param} --in {self.model_fqn} --out {self.model_fqn} '
-            f'--local '
-            f'{local_fqn} --plugin {plugin} --module {plugin} '
-            f'--lineage {self._storage_name.lineage} {resource_id}'
-        )
-        self._invoke_to_caom2(temp, plugin)
-
-    def _invoke_to_caom2(self, cmd_str, plugin):
-        """The common bits of call 'to_caom2.'"""
-        self.logger.debug(cmd_str)
-        sys.argv = cmd_str.split()
-        command = mc.load_module(plugin, 'to_caom2')
-        result = command.to_caom2()
-        if result == -1:
-            raise mc.CadcException(f'Error executing to_caom2 with {cmd_str}')
-
-    def _repo_cmd_create_client(self, observation):
-        """Create an observation instance from the input parameter."""
-        clc.repo_create(
-            self.caom_repo_client, observation, self.observable.metrics
-        )
-
-    def _repo_cmd_update_client(self, observation):
-        """Update an existing observation instance.  Assumes the obs_id
-        values are set correctly."""
-        clc.repo_update(
-            self.caom_repo_client, observation, self.observable.metrics
-        )
-
-    def _repo_cmd_read_client(self):
+    def _caom2_read(self):
         """Retrieve the existing observation model metadata."""
-        return clc.repo_get(
+        self._observation = clc.repo_get(
             self.caom_repo_client,
             self.collection,
             self._storage_name.obs_id,
             self.observable.metrics,
         )
+        self._observation_exists = False if self._observation is None else True
+
+    def _caom2_store(self):
+        """Update an existing observation instance.  Assumes the obs_id
+        values are set correctly."""
+        if self._observation_exists:
```

Comment:
Not sure `self._observation_exists` is needed when `if self._observation` achieves the same thing.

Reply:
`self._observation` cannot provide the needed information to know to call `create` or `update`. Renamed to `update_needed`, to better reflect why the variable is maintained.

If the initial observation read from the service at the beginning of an execute returns nothing, the subsequent service call needs to be `create`, otherwise it needs to be an `update`. By the time that decision needs to be made, there is always an observation to be sent to the service.